karapace/kafka_rest_apis/__init__.py

```diff
@@ -40,36 +50,70 @@ def __init__(self, config: Config) -> None:
-        self.metadata_cache = None
-        self.schemas_cache = {}
-        self.consumer_manager = ConsumerManager(config=config)
-        self.init_admin_client()
+        self.proxies: Dict[str, "UserRestProxy"] = {}
+        self._proxy_lock = asyncio.Lock()
+        log.info("REST proxy starting with (delegated authorization=%s)", self.config.get("rest_authorization", False))
+        self._idle_proxy_janitor_task: Optional[asyncio.Task] = None
 
-        self._async_producer_lock = None
-        self._async_producer: Optional[AIOKafkaProducer] = None
+    async def close(self) -> None:
+        if self._idle_proxy_janitor_task is not None:
+            self._idle_proxy_janitor_task.cancel()
+            self._idle_proxy_janitor_task = None
+        async with AsyncExitStack() as stack:
+            stack.push_async_callback(super().close)
+            stack.push_async_callback(self.serializer.close)
 
-        self.app.on_startup.append(self.initialize_asyncio_locks)
+            for proxy in self.proxies.values():
+                stack.push_async_callback(proxy.aclose)
 
-    async def initialize_asyncio_locks(self, app: aiohttp.web.Application):  # pylint: disable=unused-argument
-        """The locks can be initialized when asyncio loop is available"""
-        self.admin_lock = asyncio.Lock()
-        self._async_producer_lock = asyncio.Lock()
+    async def _idle_proxy_janitor(self) -> None:
+        IDLE_PROXY_TIMEOUT = 5 * 60
+        idle_consumer_timeout = self.config.get("consumer_idle_disconnect_timeout", 0)
+        while True:
+            await asyncio.sleep(IDLE_PROXY_TIMEOUT / 2)
+
+            key, proxy = None, None
+            async with self._proxy_lock:
+                # Always clean one at time, don't mutate dict while iterating
+                for _key, _proxy in self.proxies.items():
+                    # If UserRestProxy has consumers with state, disconnecting loses state
+                    if _proxy.num_consumers() > 0:
+                        if idle_consumer_timeout > 0 and _proxy.last_used + idle_consumer_timeout < time.monotonic():
+                            key, proxy = _key, _proxy
+                            log.warning("Disconnecting idle consumers for %s", _proxy)
+                            break
+                    # If there are no consumers, connection can be recreated without losing any state
+                    else:
+                        if _proxy.last_used + IDLE_PROXY_TIMEOUT < time.monotonic():
+                            key, proxy = _key, _proxy
+                            log.info("Releasing unused connection for %s", _proxy)
+                            break
+                if key is not None:
+                    del self.proxies[key]
+            if proxy is not None:
+                try:
+                    await proxy.aclose()
+                except:  # pylint: disable=bare-except
+                    log.exception("Disconnecting proxy failure")
```

Comment:
Log the exception?

Reply:
What do you mean?

Reply:
Ups, didn't notice it's `log.exception`, not `log.error`. All good.